File: src/java/org/voegtle/weatherstation/server/logic/WeatherDataForwarder.java
package org.voegtle.weatherstation.server.logic;

import org.json.JSONObject;
import org.voegtle.weatherstation.server.data.UnformattedWeatherDTO;
import org.voegtle.weatherstation.server.persistence.LocationProperties;
import org.voegtle.weatherstation.server.persistence.PersistenceManager;
import org.voegtle.weatherstation.server.persistence.WeatherDataSet;
import org.voegtle.weatherstation.server.util.JSONConverter;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;
import java.util.logging.Logger;

class WeatherDataForwarder {
  private static final Logger log = Logger.getLogger(WeatherDataForwarder.class.getName());
  private static final int COMMUNICATION_TIMEOUT = 60000;
  private WeatherDataFetcher weatherDataFetcher;
  private JSONConverter jsonConverter;

  WeatherDataForwarder(PersistenceManager pm, LocationProperties locationProperties) {
    weatherDataFetcher = new WeatherDataFetcher(pm, locationProperties);
    jsonConverter = new JSONConverter(locationProperties);
  }

  void forwardLastDataset() {
    UnformattedWeatherDTO latest = weatherDataFetcher.getLatestWeatherDataUnformatted(true);
    forward(jsonConverter.toJson(latest));
  }

  private void forward(JSONObject json) {
    try {
      byte encodedBytes[] = json.toString().getBytes("UTF-8");

      URL wetterCentral = new URL("https://wettercentral.appspot.com/weatherstation/cache");
      HttpURLConnection wetterConnection = (HttpURLConnection) wetterCentral.openConnection();
      wetterConnection.setConnectTimeout(COMMUNICATION_TIMEOUT);
      wetterConnection.setReadTimeout(COMMUNICATION_TIMEOUT);
      wetterConnection.setRequestMethod("POST");
      wetterConnection.setDoOutput(true);

      try {
        wetterConnection.getOutputStream().write(encodedBytes);
        wetterConnection.getOutputStream().close();
        log.info("forwarded <" + json.toString() + ">");
        String response = read(wetterConnection.getInputStream());
      } finally {
        wetterConnection.getInputStream().close();
        wetterConnection.disconnect();
      }
      
    } catch (IOException e) {
      log.warning("failed forwarding to wettercentral. " + e.getMessage());
    }
  }

  private String read(InputStream inputStream) throws IOException {
    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, "UTF-8"));
    String line = reader.readLine();
    return line;
  }

  private WeatherDataSet getLast(List<WeatherDataSet> dataSets) {
    int lastIndex = dataSets.size() - 1;
    return dataSets.get(lastIndex);
  }
}

File: src/main/java/org/voegtle/weatherstation/server/logic/WeatherDataForwarder.kt
package org.voegtle.weatherstation.server.logic

import org.springframework.web.client.RestTemplate
import org.voegtle.weatherstation.server.data.UnformattedWeatherDTO
import org.voegtle.weatherstation.server.persistence.CacheWeatherDTO
import org.voegtle.weatherstation.server.persistence.PersistenceManager
import org.voegtle.weatherstation.server.persistence.entities.LocationProperties
import java.io.IOException
import java.util.logging.Logger

internal class WeatherDataForwarder(pm: PersistenceManager, locationProperties: LocationProperties) {
    private val log = Logger.getLogger(WeatherDataForwarder::class.simpleName)

    private val weatherDataFetcher = WeatherDataFetcher(pm, locationProperties)
    private val locationProperties = locationProperties

    fun forwardLastDataset() {
        val latest = weatherDataFetcher.getLatestWeatherDataUnformatted(true)
        val cacheWeatherDTO = toCachedWeatherDTO(latest)
        forward(cacheWeatherDTO)
    }

    private fun forward(cacheWeatherDTO: CacheWeatherDTO) {
        try {
            val restTemplate = RestTemplate()
            log.info("forwarding $cacheWeatherDTO")
            restTemplate.postForObject("https://wettercentral.appspot.com/weatherstation/cache2", cacheWeatherDTO, String::class.java)
        } catch (e: IOException) {
            log.warning("failed forwarding to wettercentral. " + e.message)
        }

    }

    private fun toCachedWeatherDTO(latest: UnformattedWeatherDTO) = CacheWeatherDTO(
        id = locationProperties.location,
        time = latest.time,
        temperature = latest.temperature,
        insideTemperature = latest.insideTemperature,
        humidity = latest.humidity,
        insideHumidity = latest.insideHumidity,
        rainLastHour = if (latest.rainLastHour > 0.0f) latest.rainLastHour else null,
        rainToday = if (latest.rainToday > 0.0f) latest.rainToday else null,
        raining = latest.isRaining,
        windspeed = toKilometerPerHour(latest.windspeed),
        barometer = latest.barometer,
        solarradiation = latest.solarradiation,
        UV = latest.UV,
        powerProduction = latest.powerProduction,
        powerFeed = latest.powerFeed,
        location = latest.location,
        locationShort = locationProperties.cityShortcut,
        localTime = latest.localtime,

        forecast = locationProperties.weatherForecast,

        latitude = locationProperties.latitude,
        longitude = locationProperties.longitude
    )
}
